analyzers/src/SonarAnalyzer.Common/Extensions/SyntaxTreeExtensions.cs

```diff
@@ -18,29 +18,27 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-using System.Collections.Concurrent;
 using System.Runtime.CompilerServices;
 using Roslyn.Utilities;
 
 namespace SonarAnalyzer.Extensions;
 
 internal static class SyntaxTreeExtensions
 {
-    private static readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<SyntaxTree, bool>> GeneratedCodeCache = new();
+    private static readonly ConditionalWeakTable<SyntaxTree, object> GeneratedCodeCache = new();
 
     [PerformanceSensitive("https://github.com/SonarSource/sonar-dotnet/issues/7439", AllowCaptures = false, AllowGenericEnumeration = false, AllowImplicitBoxing = false)]
-    public static bool IsGenerated(this SyntaxTree tree, GeneratedCodeRecognizer generatedCodeRecognizer, Compilation compilation)
-    {
-        if (tree == null)
+    public static bool IsGenerated(this SyntaxTree tree, GeneratedCodeRecognizer generatedCodeRecognizer, Compilation compilation) =>
+        tree switch
         {
-            return false;
-        }
-        var cache = GeneratedCodeCache.GetOrCreateValue(compilation);
-        // Hot path: Don't use cache.GetOrAdd that takes a factory method. It allocates a delegate which causes GC pressure.
-        return cache.TryGetValue(tree, out var isGenerated)
-            ? isGenerated
-            : cache.GetOrAdd(tree, generatedCodeRecognizer.IsGenerated(tree));
-    }
+            null => false,
+            _ => GeneratedCodeCache.TryGetValue(tree, out var result)
+                ? (bool)result
+                : IsGeneratedGetOrAdd(tree, generatedCodeRecognizer)
+        };
+
+    private static bool IsGeneratedGetOrAdd(SyntaxTree tree, GeneratedCodeRecognizer generatedCodeRecognizer) =>
```

Comment:
This method can not be inlined, because of the PerformanceSensitive attribute. If we do so, The capture of `generatedCodeRecognizer` raises issues. The `IsGeneratedGetOrAdd` is no longer in the hotpath and so the PerformanceSensitive attribute doesn't apply here anymore.